src/cmd/linuxkit/azure.go

```diff
@@ -0,0 +1,372 @@
+	}
+
+	storageChannel, errorChannel := accountsClient.Create(*resourceGroup.Name, accountName, storageAccountCreateParameters, nil)
+	for {
+		select {
+		case _, ok := <-storageChannel:
+			if !ok {
+				storageChannel = nil
+			}
+		case _, ok := <-errorChannel:
+			if !ok {
+				errorChannel = nil
+			}
+		}
+		if storageChannel == nil && errorChannel == nil {
+			break
+		}
+	}
+}
+
+func uploadVMImage(resourceGroupName string, accountName string, imagePath string) {
+	accountKeys, err := accountsClient.ListKeys(resourceGroupName, accountName)
+	if err != nil {
+		fmt.Println(err.Error())
+		log.Fatalf("Unable to retrieve storage account key")
+	}
+
+	keys := *(accountKeys.Keys)
+
+	absolutePath, err := filepath.Abs(imagePath)
+	if err != nil {
+		fmt.Println(err.Error())
+		log.Fatalf("Unable to get absolute path")
+	}
+
+	directory, image := filepath.Split(absolutePath)
+
+	dockerMount := fmt.Sprintf("%s:/vhds", directory)
+	storageAccountNameArg := fmt.Sprintf("STORAGE_ACCOUNT_NAME=%s", accountName)
+	storageAccountKeyArg := fmt.Sprintf("STORAGE_ACCOUNT_KEY=%s", *keys[0].Value)
+	vhdPath := fmt.Sprintf("VHD_PATH=/vhds/%s", image)
+
+	output, err := exec.Command("docker", "run", "-v", dockerMount, "-e", vhdPath, "-e", storageAccountNameArg, "-e", storageAccountKeyArg, "radumatei/azure-vhd-upload:alpine").CombinedOutput()
+
+	if err != nil {
+		os.Stderr.WriteString(err.Error())
+	}
+	fmt.Println(string(output))
+
+	fmt.Printf("OS Image uploaded at https://%s.blob.core.windows.net/linuxkitcontainer/linuxkitimage.vhd\n", accountName)
+}
+
+func createVirtualNetwork(resourceGroup resources.Group, virtualNetworkName string, location string) *network.VirtualNetwork {
+	fmt.Printf("Creating virtual network in resource group %s, in %s", *resourceGroup.Name, location)
+
+	virtualNetworkParameters := network.VirtualNetwork{
+		Location: &location,
+		VirtualNetworkPropertiesFormat: &network.VirtualNetworkPropertiesFormat{
+			AddressSpace: &network.AddressSpace{
+				AddressPrefixes: &[]string{"10.0.0.0/16"},
+			},
+		},
+	}
+	virtualNetworkChannel, errorChannel := virtualNetworksClient.CreateOrUpdate(*resourceGroup.Name, virtualNetworkName, virtualNetworkParameters, nil)
+	var virtualNetwork network.VirtualNetwork
+	for {
+		select {
+		case v, ok := <-virtualNetworkChannel:
+			virtualNetwork = v
+			if !ok {
+				virtualNetworkChannel = nil
+			}
+		case _, ok := <-errorChannel:
+			if !ok {
+				errorChannel = nil
+			}
+		}
+		if virtualNetworkChannel == nil && errorChannel == nil {
+			break
+		}
+	}
+
+	return &virtualNetwork
+}
+
+func createSubnet(resourceGroup resources.Group, virtualNetworkName, subnetName string) *network.Subnet {
+	fmt.Printf("\nCreating subnet %s in resource group %s, within virtual network %s", subnetName, *resourceGroup.Name, virtualNetworkName)
+
+	subnetParameters := network.Subnet{
+		SubnetPropertiesFormat: &network.SubnetPropertiesFormat{
+			AddressPrefix: to.StringPtr("10.0.0.0/24"),
```

Comment:
Instead of creating all the resources (network, public IP, VM, etc) along with parameters through Go code, how about creating a JSON ARM template to create the VM and associated resources? You can instantiate the ARM template with parameters specified in JSON using the Go APIs to create a resource group and get the same result. Please see example here: https://github.com/Azure/azure-quickstart-templates/tree/master/101-vm-from-user-image

I think the ARM template will be much easier to maintain than the Go code.